### ISSUE TITLE
Pipelines fail to download datum with >10,000 files on Azure

### ISSUE THREAD

**Issue by itssimon** [pachyderm/pachyderm#3720]
**What happened?**

On a Pachyderm deployment on Azure Kubernetes Services, pipelines fail to download data if there's more than 10,000 files per datum:

```
[2019-05-11 18:20:24.553105354+10:00] starting to download data
[2019-05-11 18:29:19.311152458+10:00] errored downloading data after 8m54.47775296s: Get https://xxxxxxxxx.blob.core.windows.net/xxxxxxxxx//pach/block/xxxxxxxxx: dial tcp 13.75.240.84:443: connect: connection refused
[2019-05-11 18:29:19.314786434+10:00] failed processing datum: error downloadData: Get https://xxxxxxxxx.blob.core.windows.net/xxxxxxxxx//pach/block/xxxxxxxxx: dial tcp 13.75.240.84:443: connect: connection refused, retrying in 0s
```

The same pipeline runs fine with less than 10,000 input files.

**What you expected to happen?**

Pipeline executes successfully.

**How to reproduce it**

See https://pste.eu/p/asgi.html

**Anything else we need to know?**

While the pipelines succeeds with 8,000 files, it still takes a long time (5-10 minutes) to download the data, even though the actual data size is tiny. In our use case, this is only a small example data set, so we're planning to get this to work with a lot more files. The download time will be an issue.

**Environment?**:
- Kubernetes version: 1.14.1
- Pachyderm CLI version: 1.8.6
- Cloud provider: Azure

**Comment by itssimon** [comment]
Users in Slack (@russel) have also reported this happening in AWS with S3. I suggest implementing a backoff and retry strategy for downloading the files to prevent this. It seems that the current retry mechanism restarts the whole download from scratch and then hits the same error again.

If there is a way to download multiple files in batches rather than issuing individual GET requests, that would be ideal and would also improve the download time.

**Comment by itssimon** [comment]
A similar error also occurs when running `pachctl garbage-collect`:

```
$ pachctl garbage-collect
error deleting objects: rpc error: code = Unknown desc = Get https://pachydermstore.blob.core.windows.net/pachydermstore//pach/object/2d17a4bfecf46dc312c9cacaa1dc86cb477937a593f9535cb9ea38b052630fdca31618b2a4bb7fb63e274b55a991e0c3b4636390ed216e44bc61ce2bf50cfe4f: read tcp 10.95.8.134:60782->13.75.240.84:443: read: connection reset by peer
```

**Comment by jdoliner** [comment]
I think it's more likely that you're overwelming the network card in the VM than that the object store is throttling you, generally when object stores throttle they send back more informative errors, these look like network errors. Pachyderm does implement per-file backoff, however it only does so on errors that are on a retryable whitelist, doing so on all errors leads to errors not getting surface and turning into retry loops instead. We could consider adding the errors you're hitting to the retry list, but I'm not sure that makes sense since they can indicate a bunch of different situations, only a subset of which are temporary and thus worth retrying.

A few things worth trying here:

- larger VMs with better network cards will probably push the limit past what it is now. Probably not a long term solution, but it might get you past your current hurdle and would confirm that the network cards are really the source of slowness.
- Split the datums more finely. A datum with 10,000 files is big by pachyderm standards, if it's possible to split them into different datums with fewer files this problem should go away as they can be run in parallel on different machines.
- Combine the files in a datum together, this can be done pretty easily, by turning these files into tar archives and having your pipeline code untar them before the code runs on them.

Lastly we're currently working on a new iteration of our storage layer which will intelligently pack files into objects, so in the future, the system should naturally avoid these issues, because the number of objects requested will be based on the total data size, not the number of files.

**Comment by itssimon** [comment]
Thanks @jdoliner! I was able to resolve this issue in the meantime.

The solution was to use an Azure storage account of type "BlockBlobStorage" for the Pachyderm storage.

When following the instructions in the [Pachyderm documentation](https://docs.pachyderm.io/en/latest/deployment/azure.html), the commands create a "General-purpose V1" storage account which is actually no longer recommended and too slow for Pachyderm. Maybe it is worth updating the docs accordingly.

**Comment by itssimon** [comment]
Thanks @jdoliner! I was able resolve this in the meantime.

The trick was to use a premium storage account of type BlockBlobStorage for Pachyderm. When following the Azure deployment steps in the Pachyderm documentation it created a "General-purpose V1" storage account which is too slow (HDD based) and no longer recommended for use. I think we did actually hit object storage limits there.

Maybe it's worth updating the docs accordingly?
